File: library/src/main/java/com/pengrad/telegrambot/utility/gson/TransactionPartnerTypeAdapter.java
package com.pengrad.telegrambot.utility.gson;

import com.google.gson.*;
import com.pengrad.telegrambot.model.stars.partner.*;
import com.pengrad.telegrambot.model.stars.withdrawal.RevenueWithdrawalState;
import com.pengrad.telegrambot.model.stars.withdrawal.RevenueWithdrawalStateFailed;
import com.pengrad.telegrambot.model.stars.withdrawal.RevenueWithdrawalStatePending;
import com.pengrad.telegrambot.model.stars.withdrawal.RevenueWithdrawalStateSucceeded;

import java.lang.reflect.Type;

public class TransactionPartnerTypeAdapter implements JsonDeserializer<TransactionPartner> {

    @Override
    public TransactionPartner deserialize(JsonElement element, Type type, JsonDeserializationContext context) throws JsonParseException {
        JsonObject object = element.getAsJsonObject();
        JsonPrimitive primitive = object.getAsJsonPrimitive("type");
        String discriminator = primitive != null ? primitive.getAsString() : "unknown";

        if (TransactionPartnerUser.TYPE.equals(discriminator)) {
            return context.deserialize(object, TransactionPartnerUser.class);
        } else if (TransactionPartnerFragment.TYPE.equals(discriminator)) {
            return context.deserialize(object, TransactionPartnerFragment.class);
        } else if (TransactionPartnerTelegramAds.TYPE.equals(discriminator)) {
            return context.deserialize(object, TransactionPartnerTelegramAds.class);
        } else if (TransactionPartnerOther.TYPE.equals(discriminator)) {
            return context.deserialize(object, TransactionPartnerOther.class);
        }

        return new TransactionPartner(discriminator);
    }
}

File: library/src/main/java/com/pengrad/telegrambot/utility/gson/TransactionPartnerTypeAdapter.kt
package com.pengrad.telegrambot.utility.gson

import com.google.gson.*
import com.pengrad.telegrambot.model.stars.partner.*
import com.pengrad.telegrambot.model.stars.partner.TransactionPartnerType.AFFILIATE_PROGRAM
import com.pengrad.telegrambot.model.stars.partner.TransactionPartnerType.FRAGMENT
import com.pengrad.telegrambot.model.stars.partner.TransactionPartnerType.OTHER
import com.pengrad.telegrambot.model.stars.partner.TransactionPartnerType.TELEGRAM_ADS
import com.pengrad.telegrambot.model.stars.partner.TransactionPartnerType.TELEGRAM_API
import com.pengrad.telegrambot.model.stars.partner.TransactionPartnerType.UNKNOWN
import com.pengrad.telegrambot.model.stars.partner.TransactionPartnerType.USER
import java.lang.reflect.Type

object TransactionPartnerTypeAdapter : JsonDeserializer<TransactionPartner> {

    private val typeMapping = mapOf(
        USER to TransactionPartnerUser::class,
        AFFILIATE_PROGRAM to TransactionPartnerAffiliateProgram::class,
        FRAGMENT to TransactionPartnerFragment::class,
        TELEGRAM_ADS to TransactionPartnerTelegramAds::class,
        TELEGRAM_API to TransactionPartnerTelegramApi::class,
        OTHER to TransactionPartnerOther::class
    )

    @Throws(JsonParseException::class)
    override fun deserialize(
        element: JsonElement,
        type: Type,
        context: JsonDeserializationContext
    ): TransactionPartner {
        val obj = element.asJsonObject
        val discriminator = obj.getAsJsonPrimitive("type")?.asString ?: UNKNOWN

        return typeMapping[discriminator]?.let {
            context.deserialize(obj, it.java)
        } ?: TransactionPartnerUnknown(discriminator)
    }

}